utilities/blob_db/blob_db_impl.cc

```diff
@@ -0,0 +1,2112 @@
+    const std::vector<Slice>& keys, std::vector<std::string>* values) {
+  auto statuses = db_->MultiGet(options, column_family, keys, values);
+
+  for (size_t i = 0; i < keys.size(); ++i) {
+    if (!statuses[i].ok()) continue;
+
+    BlockHandle handle;
+    Slice index_entry_slice((*values)[i]);
+    uint64_t file_number;
+    if (!GetVarint64(&index_entry_slice, &file_number)) {
+      statuses[i] = Status::Corruption();
+      continue;
+    }
+
+    statuses[i] = handle.DecodeFrom(&index_entry_slice);
+    if (!statuses[i].ok()) continue;
+
+    std::shared_ptr<BlobFile> bfile;
+    {
+      ReadLock l(&mutex_);
+      auto hitr = blob_files_.find(file_number);
+
+      // file was deleted
+      if (hitr == blob_files_.end()) {
+        statuses[i] = Status::NotFound("Blob Not Found");
+        continue;
+      }
+
+      bfile = hitr->second;
+    }
+
+    std::shared_ptr<RandomAccessFileReader> reader;
+    {
+      WriteLock lockbfile_w(&bfile->mutex_);
+      reader = openRandomAccess_locked(bfile, myenv_, env_options_);
+    }
+
+    // allocate the buffer
+    (*values)[i].resize(handle.size());
+    char* buffer = &((*values)[i])[0];
+
+    Slice blob_value;
+    statuses[i] =
+        reader->Read(handle.offset(), handle.size(), &blob_value, buffer);
+  }
+  return statuses;
+}
+
+Status BlobDBImpl::Get(const ReadOptions& options,
+                       ColumnFamilyHandle* column_family, const Slice& key,
+                       std::string* value) {
+  Status s;
+  std::string index_entry;
+  s = db_->Get(options, column_family, key, &index_entry);
+  if (!s.ok()) {
+    if (debug_level_ >= 3)
+      Log(InfoLogLevel::WARN_LEVEL, db_options_.info_log,
+          "Get Failed on LSM KEY: %s status: '%s'", key.ToString().c_str(),
+          s.ToString().c_str());
+    return s;
+  }
+
+  BlockHandle handle;
+  Slice index_entry_slice(index_entry);
+  uint64_t file_number;
+  if (!GetVarint64(&index_entry_slice, &file_number))
+    return Status::Corruption();
+
+  s = handle.DecodeFrom(&index_entry_slice);
+  if (!s.ok()) return s;
+
+  // offset has to have certain min, as we will read CRC
+  // later from the Blob Header, which needs to be also a
+  // valid offset.
+  if (handle.offset() <
+      (BlobLogHeader::kHeaderSize + BlobLogRecord::kHeaderSize + key.size())) {
+    if (debug_level_ >= 2) {
+      Log(InfoLogLevel::ERROR_LEVEL, db_options_.info_log,
+          "Invalid blob handle file_number: %" PRIu64 " blob_offset: %" PRIu64
+          " blob_size: %" PRIu64 " key: %s",
+          file_number, handle.offset(), handle.size(), key.data());
+    }
+    return Status::NotFound("Blob Not Found, although found in LSM");
+  }
+
+  std::shared_ptr<BlobFile> bfile;
+  {
+    ReadLock l(&mutex_);
+    auto hitr = blob_files_.find(file_number);
+
+    // file was deleted
+    if (hitr == blob_files_.end()) {
+      return Status::NotFound("Blob Not Found as blob file missing");
```

Comment:
I think I have asked this before: how do you deal with the case that the blob file is compacted away after we fetched the index?